esque/ruleparser/expressionelement.py

```diff
@@ -0,0 +1,220 @@
+
+    FIELDS = {
+        "MESSAGE_KEY": "message\\.key",
+        "MESSAGE_LENGTH": "message\\.length",
+        "MESSAGE_PARTITION": "message\\.partition",
+        "MESSAGE_HEADER": "message\\.header\\.[a-zA-Z0-9_]+",
+        "MESSAGE_OFFSET": "message\\.offset",
+        "MESSAGE_TIMESTAMP": "message\\.timestamp",
+        "SYSTEM_TIMESTAMP": "system\\.timestamp",
+    }
+    OPERATORS = {
+        "ADDITION": GenericOperator("ADDITION", OperatorType.ARITHMETIC, "\\+", 20, 20, False),
+        "SUBTRACTION": GenericOperator("SUBTRACTION", OperatorType.ARITHMETIC, "-", 20, 20, False),
+        "MULTIPLICATION": GenericOperator("MULTIPLICATION", OperatorType.ARITHMETIC, "\\*", 30, 30, False),
+        "DIVISION": GenericOperator("DIVISION", OperatorType.ARITHMETIC, "\\/", 30, 30, False),
+        "REMAINDER": GenericOperator("REMAINDER", OperatorType.ARITHMETIC, "mod", 20, 20, False),
+        "BINARY_AND": GenericOperator("BINARY_AND", OperatorType.ARITHMETIC, "&", 20, 20, False),
+        "BINARY_OR": GenericOperator("BINARY_OR", OperatorType.ARITHMETIC, "\\|", 20, 20, False),
+        "BINARY_XOR": GenericOperator("BINARY_XOR", OperatorType.ARITHMETIC, "\\^", 20, 20, False),
+        "GREATER_THAN": GenericOperator("GREATER_THAN", OperatorType.COMPARISON, ">", 15, 15, False),
+        "LESS_THAN": GenericOperator("LESS_THAN", OperatorType.COMPARISON, "<", 15, 15, False),
+        "GREATER_OR_EQUAL": GenericOperator("GREATER_OR_EQUAL", OperatorType.COMPARISON, ">=", 15, 15, False),
+        "LESS_OR_EQUAL": GenericOperator("LESS_OR_EQUAL", OperatorType.COMPARISON, "<=", 15, 15, False),
+        "EQUAL": GenericOperator("EQUAL", OperatorType.COMPARISON, "==", 15, 15, False),
+        "NOT_EQUAL": GenericOperator("NOT_EQUAL", OperatorType.COMPARISON, "!=", 15, 15, False),
+        "LOGICAL_AND": GenericOperator("LOGICAL_AND", OperatorType.ARITHMETIC, "and", 10, 10, False),
+        "LOGICAL_OR": GenericOperator("LOGICAL_OR", OperatorType.ARITHMETIC, "or", 10, 10, False),
+        "LOGICAL_NOT": GenericOperator("LOGICAL_NOT", OperatorType.COMPARISON, "neg", 100, 100, True),
+        "LOGICAL_XOR": GenericOperator("LOGICAL_XOR", OperatorType.ARITHMETIC, "xor", 10, 10, False),
+        "LIKE": GenericOperator("LIKE", OperatorType.COMPARISON, "like", 15, 15, False),
+        "NOT_LIKE": GenericOperator("NOT_LIKE", OperatorType.COMPARISON, "notlike", 15, 15, False),
+        "PARENTHESIS_OPEN": GenericOperator("PARENTHESIS_OPEN", OperatorType.PARENTHESIS, "\\(", 10000, 0, False),
+        "PARENTHESIS_CLOSED": GenericOperator("PARENTHESIS_CLOSED", OperatorType.PARENTHESIS, "\\)", 1, 0, False),
+    }
+
+
+class AbstractBinaryOperator(Operator):
+    def evaluate(self, operand1, operand2):
+        pass
+
+
+class AbstractUnaryOperator(Operator):
+    def evaluate(self, operand1):
+        pass
+
+
+class ArithmeticBinaryOperator(AbstractBinaryOperator):
+    def evaluate(self, operand1: str, operand2: str):
+        op1_converted = None
```

Comment:
the first part of the evaluate function is the same for ArithmeticBinaryOperator and ComparisonBinaryOperator, how about pulling them out into a function in AbstractBinaryOperator called OperandTypeConversions or something?

Reply:
Moved the evaluation to the tuple defining the operators.